### DISCUSSION TITLE
Odd behaviour with Parameters injected into Services injected into Controller actions

### DISCUSSION THREAD

**Discussion by HenryDeWitt1** [symfony/symfony#48111] [Q&A]
I can't find anything online related to this so thought i'd ask here. 

On Symfony 4.4.48 I have a service which requires a few parameters, these are bound and injected in the config yml and autowired into the service constructor, this service is then injected into a controller action. In a recent change I accidently deleted one of the parameters from the bind config but the container continues to build with no exception thrown.

However when checking the cache the file related to the service it contains a throw referencing the missing parameter, so the build process has identified that this parameter is missing but is not actually throwing the exception at build time. If i move the service injection from the controller action to the controller constructor then the exception is thrown as you would expect.

I can see how the service being injected into the action instead of the constructor means that it technically isn't needed for the controller as a service to run but also means that its only at runtime when the service is called that an exception is shown.

So my question, is this expected behaviour?

**Comment by MatTheCat** [comment]
> I accidently deleted one of the parameters from the bind config

Do you mean you removed the binding or the parameter?

**Reply by HenryDeWitt1** [reply]
I removed the binding

**Reply by MatTheCat** [reply]
Okay so yes this is expected: services injected in controllers actions end up in services locators to make them lazy, so errors only pop when they’re actually used.

**Reply by HenryDeWitt1** [reply]
I thought that might be the case, seems a bit inconsistent though, especially when the missing dependency is recognised at compile time.

So to me it looks like there is a choice between optimisation and reliability, in terms for reliability and avoiding these issues showing up at runtime its not possible to use the action level injection and services need to be injected into the constructor of the controller. However this means that all dependencies of a controller are instantiated even if they are not used by a given action servicing the request.

**Reply by MatTheCat** [reply]
As you’re supposed to test your application I don’t think this is a problem, but you can still open an issue about it :thinking:

**Reply by HenryDeWitt1** [reply]
Thanks for your help. This issue was discovered because I added a new service with a bound parameter then merged develop and somehow missed that one of the existing binds was removed by the conflict resolution. It just so happened the dev that added that parameter was my reviewer and had added said parameter last week which is why it was noticed at all. Although his ticket has passed QA, our automation team have not had time to add automation testing for his feature yet so the bug would not have shown up for some time.

I think it would be better if the throw message caused by the missing dependency were to show up at compile time instead of being contingent on which of the standard injection options are used in controllers.

**Reply by MatTheCat** [reply]
Maybe it’s feasible; I think you should open an issue: in the worst case you’ll know why it’s not.

May be linked to `RUNTIME_EXCEPTION_ON_INVALID_REFERENCE` use in `RegisterControllerArgumentLocatorsPass`.

**Reply by HenryDeWitt1** [reply]
Thanks I have raised a bug. 
I did some further testing and added a second service:
1. Controller->service1->parameter
2. Controller->service2->service1->parameter
With this setup (when injecting into the action) the missing bind is flagged at compile time for 2 but not for 1. So it looks like this is only an issue if the service directly injected into the action is missing a dependency.
